File: Frontend/lib/models/emotion.dart
import 'user.dart';

class Emotion {
  final String id;
  final User user;
  final OriginalImage originalImage;
  final DetectedEmotion detectedEmotion;
  final GhibliArt ghibliArt;
  final EmotionMetadata metadata;
  final EmotionSharing sharing;
  final List<EmotionReaction> reactions;
  final Map<String, int> reactionCounts;
  final bool isDeleted;
  final DateTime? deletedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  Emotion({
    required this.id,
    required this.user,
    required this.originalImage,
    required this.detectedEmotion,
    required this.ghibliArt,
    required this.metadata,
    required this.sharing,
    required this.reactions,
    required this.reactionCounts,
    required this.isDeleted,
    this.deletedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Emotion.fromJson(Map<String, dynamic> json) {
    return Emotion(
      id: json['_id'] ?? json['id'],
      user: User.fromJson(json['user']),
      originalImage: OriginalImage.fromJson(json['originalImage']),
      detectedEmotion: DetectedEmotion.fromJson(json['detectedEmotion']),
      ghibliArt: GhibliArt.fromJson(json['ghibliArt']),
      metadata: EmotionMetadata.fromJson(json['metadata']),
      sharing: EmotionSharing.fromJson(json['sharing']),
      reactions: (json['reactions'] as List<dynamic>?)
          ?.map((e) => EmotionReaction.fromJson(e))
          .toList() ?? [],
      reactionCounts: Map<String, int>.from(json['reactionCounts'] ?? {}),
      isDeleted: json['isDeleted'] ?? false,
      deletedAt: json['deletedAt'] != null ? DateTime.parse(json['deletedAt']) : null,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'user': user.toJson(),
      'originalImage': originalImage.toJson(),
      'detectedEmotion': detectedEmotion.toJson(),
      'ghibliArt': ghibliArt.toJson(),
      'metadata': metadata.toJson(),
      'sharing': sharing.toJson(),
      'reactions': reactions.map((e) => e.toJson()).toList(),
      'reactionCounts': reactionCounts,
      'isDeleted': isDeleted,
      'deletedAt': deletedAt?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class OriginalImage {
  final String url;
  final String filename;
  final int size;
  final String mimeType;

  OriginalImage({
    required this.url,
    required this.filename,
    required this.size,
    required this.mimeType,
  });

  factory OriginalImage.fromJson(Map<String, dynamic> json) {
    return OriginalImage(
      url: json['url'],
      filename: json['filename'],
      size: json['size'],
      mimeType: json['mimeType'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'url': url,
      'filename': filename,
      'size': size,
      'mimeType': mimeType,
    };
  }
}

class DetectedEmotion {
  final String primary;
  final double confidence;
  final List<EmotionResult> allEmotions;

  DetectedEmotion({
    required this.primary,
    required this.confidence,
    required this.allEmotions,
  });

  factory DetectedEmotion.fromJson(Map<String, dynamic> json) {
    return DetectedEmotion(
      primary: json['primary'],
      confidence: json['confidence'].toDouble(),
      allEmotions: (json['allEmotions'] as List<dynamic>)
          .map((e) => EmotionResult.fromJson(e))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'primary': primary,
      'confidence': confidence,
      'allEmotions': allEmotions.map((e) => e.toJson()).toList(),
    };
  }
}

class EmotionResult {
  final String emotion;
  final double confidence;

  EmotionResult({
    required this.emotion,
    required this.confidence,
  });

  factory EmotionResult.fromJson(Map<String, dynamic> json) {
    return EmotionResult(
      emotion: json['emotion'],
      confidence: json['confidence'].toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'emotion': emotion,
      'confidence': confidence,
    };
  }
}

class GhibliArt {
  final String url;
  final String style;
  final String prompt;
  final double generationTime;

  GhibliArt({
    required this.url,
    required this.style,
    required this.prompt,
    required this.generationTime,
  });

  factory GhibliArt.fromJson(Map<String, dynamic> json) {
    return GhibliArt(
      url: json['url'],
      style: json['style'],
      prompt: json['prompt'],
      generationTime: json['generationTime'].toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'url': url,
      'style': style,
      'prompt': prompt,
      'generationTime': generationTime,
    };
  }
}

class EmotionMetadata {
  final bool faceDetected;
  final int faceCount;
  final String imageQuality;
  final double processingTime;
  final String apiVersion;

  EmotionMetadata({
    required this.faceDetected,
    required this.faceCount,
    required this.imageQuality,
    required this.processingTime,
    required this.apiVersion,
  });

  factory EmotionMetadata.fromJson(Map<String, dynamic> json) {
    return EmotionMetadata(
      faceDetected: json['faceDetected'],
      faceCount: json['faceCount'],
      imageQuality: json['imageQuality'],
      processingTime: json['processingTime'].toDouble(),
      apiVersion: json['apiVersion'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'faceDetected': faceDetected,
      'faceCount': faceCount,
      'imageQuality': imageQuality,
      'processingTime': processingTime,
      'apiVersion': apiVersion,
    };
  }
}

class EmotionSharing {
  final bool isShared;
  final List<SharedWith> sharedWith;
  final DateTime? sharedAt;

  EmotionSharing({
    required this.isShared,
    required this.sharedWith,
    this.sharedAt,
  });

  factory EmotionSharing.fromJson(Map<String, dynamic> json) {
    return EmotionSharing(
      isShared: json['isShared'] ?? false,
      sharedWith: (json['sharedWith'] as List<dynamic>?)
          ?.map((e) => SharedWith.fromJson(e))
          .toList() ?? [],
      sharedAt: json['sharedAt'] != null ? DateTime.parse(json['sharedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'isShared': isShared,
      'sharedWith': sharedWith.map((e) => e.toJson()).toList(),
      'sharedAt': sharedAt?.toIso8601String(),
    };
  }
}

class SharedWith {
  final User user;
  final String? message;
  final DateTime sharedAt;

  SharedWith({
    required this.user,
    this.message,
    required this.sharedAt,
  });

  factory SharedWith.fromJson(Map<String, dynamic> json) {
    return SharedWith(
      user: User.fromJson(json['user']),
      message: json['message'],
      sharedAt: DateTime.parse(json['sharedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user': user.toJson(),
      'message': message,
      'sharedAt': sharedAt.toIso8601String(),
    };
  }
}

class EmotionReaction {
  final User user;
  final String reaction;
  final DateTime reactedAt;

  EmotionReaction({
    required this.user,
    required this.reaction,
    required this.reactedAt,
  });

  factory EmotionReaction.fromJson(Map<String, dynamic> json) {
    return EmotionReaction(
      user: User.fromJson(json['user']),
      reaction: json['reaction'],
      reactedAt: DateTime.parse(json['reactedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user': user.toJson(),
      'reaction': reaction,
      'reactedAt': reactedAt.toIso8601String(),
    };
  }
}

File: Backend/models/Emotion.js
const mongoose = require('mongoose');

const emotionSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  originalImage: {
    url: {
      type: String,
      required: true
    },
    publicId: String, // For Cloudinary
    filename: String,
    size: Number,
    mimeType: String
  },
  detectedEmotion: {
    primary: {
      type: String,
      required: true,
      enum: ['happy', 'sad', 'angry', 'surprised', 'neutral', 'excited', 'confused', 'disgusted']
    },
    confidence: {
      type: Number,
      required: true,
      min: 0,
      max: 1
    },
    allEmotions: [{
      emotion: {
        type: String,
        enum: ['happy', 'sad', 'angry', 'surprised', 'neutral', 'excited', 'confused', 'disgusted']
      },
      confidence: {
        type: Number,
        min: 0,
        max: 1
      }
    }]
  },
  ghibliArt: {
    url: {
      type: String,
      required: true
    },
    publicId: String, // For Cloudinary
    style: {
      type: String,
      default: 'ghibli'
    },
    prompt: String, // The prompt used to generate the art
    generationTime: Number // Time taken to generate in seconds
  },
  metadata: {
    faceDetected: {
      type: Boolean,
      default: true
    },
    faceCount: {
      type: Number,
      default: 1
    },
    imageQuality: {
      type: String,
      enum: ['low', 'medium', 'high'],
      default: 'medium'
    },
    processingTime: Number, // Total processing time in seconds
    apiVersion: String // Version of the emotion detection API used
  },
  sharing: {
    isShared: {
      type: Boolean,
      default: false
    },
    sharedWith: [{
      user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User'
      },
      sharedAt: {
        type: Date,
        default: Date.now
      },
      message: String // Optional message when sharing
    }],
    isPublic: {
      type: Boolean,
      default: false
    },
    allowDownload: {
      type: Boolean,
      default: false
    }
  },
  reactions: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    reaction: {
      type: String,
      enum: ['like', 'love', 'laugh', 'wow', 'sad', 'angry']
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  tags: [{
    type: String,
    trim: true,
    lowercase: true
  }],
  isDeleted: {
    type: Boolean,
    default: false
  },
  deletedAt: Date,
  expiresAt: {
    type: Date,
    default: function() {
      // Auto-delete after 30 days if not shared
      return new Date(Date.now() + 30 * 24 * 60 * 60 * 1000);
    }
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Virtual for reaction counts
emotionSchema.virtual('reactionCounts').get(function() {
  const counts = {};
  this.reactions.forEach(reaction => {
    counts[reaction.reaction] = (counts[reaction.reaction] || 0) + 1;
  });
  return counts;
});

// Virtual for total shares
emotionSchema.virtual('shareCount').get(function() {
  return this.sharing.sharedWith ? this.sharing.sharedWith.length : 0;
});

// Index for better performance
emotionSchema.index({ user: 1, createdAt: -1 });
emotionSchema.index({ 'detectedEmotion.primary': 1 });
emotionSchema.index({ 'sharing.isPublic': 1 });
emotionSchema.index({ 'sharing.isShared': 1 });
emotionSchema.index({ tags: 1 });
emotionSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });

// Pre-save middleware
emotionSchema.pre('save', function(next) {
  // If emotion is shared, remove expiration
  if (this.sharing.isShared && this.expiresAt) {
    this.expiresAt = undefined;
  }
  next();
});

// Method to share with user
emotionSchema.methods.shareWith = function(userId, message = '') {
  // Check if already shared with this user
  const existingShare = this.sharing.sharedWith.find(share => 
    share.user.toString() === userId.toString()
  );
  
  if (!existingShare) {
    this.sharing.sharedWith.push({
      user: userId,
      sharedAt: new Date(),
      message: message
    });
    this.sharing.isShared = true;
    // Remove expiration when shared
    this.expiresAt = undefined;
  }
};

// Method to unshare with user
emotionSchema.methods.unshareWith = function(userId) {
  this.sharing.sharedWith = this.sharing.sharedWith.filter(share => 
    share.user.toString() !== userId.toString()
  );
  
  // If no more shares, mark as not shared
  if (this.sharing.sharedWith.length === 0) {
    this.sharing.isShared = false;
    // Reset expiration
    this.expiresAt = new Date(Date.now() + 30 * 24 * 60 * 60 * 1000);
  }
};

// Method to add reaction
emotionSchema.methods.addReaction = function(userId, reactionType) {
  // Remove existing reaction from this user
  this.reactions = this.reactions.filter(reaction => 
    reaction.user.toString() !== userId.toString()
  );
  
  // Add new reaction
  this.reactions.push({
    user: userId,
    reaction: reactionType,
    createdAt: new Date()
  });
};

// Method to remove reaction
emotionSchema.methods.removeReaction = function(userId) {
  this.reactions = this.reactions.filter(reaction => 
    reaction.user.toString() !== userId.toString()
  );
};

// Method to check if user can access
emotionSchema.methods.canAccess = function(userId) {
  // Owner can always access
  if (this.user.toString() === userId.toString()) {
    return true;
  }
  
  // Check if shared with user
  if (this.sharing.isShared) {
    return this.sharing.sharedWith.some(share => 
      share.user.toString() === userId.toString()
    ) || this.sharing.isPublic;
  }
  
  return false;
};

// Static method to get emotions for user
emotionSchema.statics.getEmotionsForUser = function(userId, page = 1, limit = 20) {
  const skip = (page - 1) * limit;
  
  return this.find({
    user: userId,
    isDeleted: false
  })
  .sort({ createdAt: -1 })
  .skip(skip)
  .limit(limit)
  .populate('user', 'name email avatar');
};

// Static method to get shared emotions for user
emotionSchema.statics.getSharedEmotionsForUser = function(userId, page = 1, limit = 20) {
  const skip = (page - 1) * limit;
  
  return this.find({
    $or: [
      { 'sharing.sharedWith.user': userId },
      { 'sharing.isPublic': true }
    ],
    isDeleted: false
  })
  .sort({ createdAt: -1 })
  .skip(skip)
  .limit(limit)
  .populate('user', 'name email avatar');
};

// Static method to get emotion statistics for user
emotionSchema.statics.getEmotionStatsForUser = function(userId, days = 30) {
  const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000);
  
  return this.aggregate([
    {
      $match: {
        user: userId,
        createdAt: { $gte: startDate },
        isDeleted: false
      }
    },
    {
      $group: {
        _id: '$detectedEmotion.primary',
        count: { $sum: 1 },
        avgConfidence: { $avg: '$detectedEmotion.confidence' }
      }
    },
    {
      $sort: { count: -1 }
    }
  ]);
};

module.exports = mongoose.model('Emotion', emotionSchema);